File: app/src/main/java/com/zacharee1/systemuituner/activites/apppickers/ImmersiveSelectActivity.java
package com.zacharee1.systemuituner.activites.apppickers;

import android.app.Fragment;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Looper;
import android.preference.CheckBoxPreference;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.zacharee1.systemuituner.R;
import com.zacharee1.systemuituner.misc.AppInfo;
import com.zacharee1.systemuituner.handlers.ImmersiveHandler;
import com.zacharee1.systemuituner.util.Utils;

import java.util.ArrayList;
import java.util.TreeMap;
import java.util.TreeSet;

public class ImmersiveSelectActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setTheme(Utils.isInDarkMode(this) ? R.style.AppTheme_Dark_NoActionBar : R.style.AppTheme_NoActionBar);

        setContentView(R.layout.activity_blank_custom_toolbar);

        final ProgressBar bar = new ProgressBar(this);
        bar.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));

        ((LinearLayout)findViewById(R.id.content_main)).addView(bar);

        new Thread(new Runnable() {
            @Override
            public void run() {
                Looper.prepare();

                final TreeMap<String, AppInfo> appMap = new TreeMap<>();

                for (ApplicationInfo info : Utils.getInstalledApps(ImmersiveSelectActivity.this)) {
                    try {
                        if (getPackageManager().getPackageInfo(info.packageName, PackageManager.GET_ACTIVITIES).activities.length > 1) {
                            appMap.put(info.loadLabel(getPackageManager()).toString(),
                                    new AppInfo(info.loadLabel(getPackageManager()).toString(),
                                            info.packageName,
                                            null,
                                            info.loadIcon(getPackageManager()))
                            );
                        }
                    } catch (Exception e) {}
                }

                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        final SelectorFragment fragment = SelectorFragment.newInstance();
                        fragment.setInfo(appMap);

                        ((LinearLayout)findViewById(R.id.content_main)).removeAllViews();
                        try {
                            getFragmentManager().beginTransaction().replace(R.id.content_main, fragment).commit();
                        } catch (Exception e) {}
                        setUpActionBar(fragment);
                    }
                });
            }
        }).start();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void setUpActionBar(final SelectorFragment fragment) {
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        View selectAll = LayoutInflater.from(this).inflate(R.layout.select_all, toolbar, false);
        View deselectAll = LayoutInflater.from(this).inflate(R.layout.deselect_all, toolbar, false);
        View invertSelection = LayoutInflater.from(this).inflate(R.layout.invert_select, toolbar, false);

        toolbar.addView(selectAll);
        toolbar.addView(deselectAll);
        toolbar.addView(invertSelection);

        selectAll.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                fragment.selectAllBoxes();
            }
        });
        deselectAll.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                fragment.deselectAllBoxes();
            }
        });
        invertSelection.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                fragment.invertSelection();
            }
        });

        selectAll.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                Toast.makeText(ImmersiveSelectActivity.this,
                        getResources().getString(R.string.select_all),
                        Toast.LENGTH_SHORT)
                        .show();
                return true;
            }
        });
        deselectAll.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                Toast.makeText(ImmersiveSelectActivity.this,
                        getResources().getString(R.string.deselect_all),
                        Toast.LENGTH_SHORT)
                        .show();
                return true;
            }
        });
        invertSelection.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                Toast.makeText(ImmersiveSelectActivity.this,
                        getResources().getString(R.string.invert_selection),
                        Toast.LENGTH_SHORT)
                        .show();
                return true;
            }
        });
    }

    public static class SelectorFragment extends PreferenceFragment {
        TreeMap<String, AppInfo> mInfo;

        public static SelectorFragment newInstance() {
            return new SelectorFragment();
        }

        public void setInfo(TreeMap<String, AppInfo> info) {
            mInfo = info;
        }

        @Override
        public void onCreate(@Nullable Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);

            addPreferencesFromResource(R.xml.pref_blank);
            populateList();
        }

        public void populateList() {
            TreeSet<String> selectedApps = ImmersiveHandler.parseSelectedApps(getActivity(), new TreeSet<String>());

            for (AppInfo info : mInfo.values()) {
                CheckBoxPreference preference = new CheckBoxPreference(getActivity());
                preference.setTitle(info.appName);
                preference.setSummary(info.packageName);
                preference.setIcon(info.appIcon);
                preference.setKey(info.packageName);
                preference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
                    @Override
                    public boolean onPreferenceChange(Preference preference, Object o) {
                        boolean isChecked = Boolean.valueOf(o.toString());
                        if (isChecked) {
                            ImmersiveHandler.addApp(getActivity(), preference.getKey());
                        } else {
                            ImmersiveHandler.removeApp(getActivity(), preference.getKey());
                        }
                        return true;
                    }
                });
                preference.setChecked(selectedApps.contains(preference.getKey()));

                getPreferenceScreen().addPreference(preference);
            }
        }

        public void selectAllBoxes() {
            setBoxesSelected(true);
        }

        public void deselectAllBoxes() {
            setBoxesSelected(false);
        }

        public void setBoxesSelected(boolean selected) {
            for (int i = 0; i < getPreferenceScreen().getPreferenceCount(); i++) {
                Preference p = getPreferenceScreen().getPreference(i);

                if (p instanceof CheckBoxPreference) {
                    ((CheckBoxPreference) p).setChecked(selected);
                }
            }
        }

        public void invertSelection() {
            ArrayList<CheckBoxPreference> selected = new ArrayList<>();
            ArrayList<CheckBoxPreference> unselected = new ArrayList<>();

            for (int i = 0; i < getPreferenceScreen().getPreferenceCount(); i++) {
                Preference p = getPreferenceScreen().getPreference(i);

                if (p instanceof CheckBoxPreference) {
                    CheckBoxPreference box = (CheckBoxPreference) p;
                    if (box.isChecked()) {
                        selected.add(box);
                    } else {
                        unselected.add(box);
                    }
                }
            }

            for (CheckBoxPreference box : selected) {
                box.setChecked(false);
            }

            for (CheckBoxPreference box : unselected) {
                box.setChecked(true);
            }
        }
    }
}

File: app/src/main/java/com/zacharee1/systemuituner/activites/apppickers/ImmersiveSelectActivity.kt
package com.zacharee1.systemuituner.activites.apppickers

import android.content.pm.PackageManager
import android.os.Bundle
import android.preference.CheckBoxPreference
import android.preference.Preference
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import android.widget.Toast
import com.dinuscxj.progressbar.CircleProgressBar
import com.zacharee1.systemuituner.R
import com.zacharee1.systemuituner.activites.BaseAnimActivity
import com.zacharee1.systemuituner.fragments.AnimFragment
import com.zacharee1.systemuituner.handlers.ImmersiveHandler
import com.zacharee1.systemuituner.misc.AppInfo
import com.zacharee1.systemuituner.util.getInstalledApps
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import java.util.*

class ImmersiveSelectActivity : BaseAnimActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.immersive_select)
        setTitle(R.string.select_apps)

        val bar = findViewById<CircleProgressBar>(R.id.app_load_progress)

        Observable.fromCallable { getInstalledApps() }
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .subscribe {
                    val appMap = TreeMap<String, AppInfo>()

                    it.forEach { info ->
                        val activities = packageManager.getPackageInfo(info.packageName, PackageManager.GET_ACTIVITIES).activities
                        if (activities?.isNotEmpty() == true) {
                            appMap[info.loadLabel(packageManager).toString()] = AppInfo(info.loadLabel(packageManager).toString(),
                                    info.packageName,
                                    null,
                                    info.loadIcon(packageManager))
                            runOnUiThread { bar.progress = 100 * (it.indexOf(info) + 1) / it.size }
                        }
                    }

                    runOnUiThread {
                        val fragment = SelectorFragment.newInstance()
                        fragment.setInfo(appMap)

                        (findViewById<View>(R.id.content_main) as LinearLayout).removeAllViews()
                        try {
                            fragmentManager.beginTransaction().replace(R.id.content_main, fragment).commit()
                        } catch (e: Exception) {
                        }

                        setUpActionBar(fragment)
                    }
                }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                finish()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun setUpActionBar(fragment: SelectorFragment) {
        val selectAll = LayoutInflater.from(this).inflate(R.layout.select_all, toolbar, false)
        val deselectAll = LayoutInflater.from(this).inflate(R.layout.deselect_all, toolbar, false)
        val invertSelection = LayoutInflater.from(this).inflate(R.layout.invert_select, toolbar, false)

        toolbar.addView(selectAll)
        toolbar.addView(deselectAll)
        toolbar.addView(invertSelection)

        selectAll.setOnClickListener { fragment.selectAllBoxes() }
        deselectAll.setOnClickListener { fragment.deselectAllBoxes() }
        invertSelection.setOnClickListener { fragment.invertSelection() }

        selectAll.setOnLongClickListener {
            Toast.makeText(this@ImmersiveSelectActivity,
                    resources.getString(R.string.select_all),
                    Toast.LENGTH_SHORT)
                    .show()
            true
        }
        deselectAll.setOnLongClickListener {
            Toast.makeText(this@ImmersiveSelectActivity,
                    resources.getString(R.string.deselect_all),
                    Toast.LENGTH_SHORT)
                    .show()
            true
        }
        invertSelection.setOnLongClickListener {
            Toast.makeText(this@ImmersiveSelectActivity,
                    resources.getString(R.string.invert_selection),
                    Toast.LENGTH_SHORT)
                    .show()
            true
        }
    }

    class SelectorFragment : AnimFragment() {
        private var infos: TreeMap<String, AppInfo> = TreeMap()

        fun setInfo(info: TreeMap<String, AppInfo>) {
            infos.putAll(info)
        }

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)

            addPreferencesFromResource(R.xml.pref_blank)
            populateList()
        }

        override fun onSetTitle(): String? = resources.getString(R.string.select_apps)

        private fun populateList() {
            val selectedApps = ImmersiveHandler.parseSelectedApps(activity, TreeSet())

            for (info in infos.values) {
                val preference = CheckBoxPreference(activity)
                preference.title = info.appName
                preference.summary = info.packageName
                preference.icon = info.appIcon
                preference.key = info.packageName
                preference.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _, o ->
                    if (activity != null) {
                        val isChecked = o.toString().toBoolean()
                        if (isChecked) {
                            ImmersiveHandler.addApp(activity, preference.key)
                        } else {
                            ImmersiveHandler.removeApp(activity, preference.key)
                        }
                        restartMode()
                        true
                    } else false
                }
                preference.isChecked = selectedApps.contains(preference.key)

                preferenceScreen.addPreference(preference)
            }
        }

        fun selectAllBoxes() {
            setBoxesSelected(true)
        }

        fun deselectAllBoxes() {
            setBoxesSelected(false)
        }

        private fun setBoxesSelected(selected: Boolean) {
            (0 until preferenceScreen.preferenceCount)
                    .map { preferenceScreen.getPreference(it) }
                    .filterIsInstance<CheckBoxPreference>()
                    .apply {
                        forEach {
                            it.isChecked = selected
                        }
                    }
        }

        fun invertSelection() {
            (0 until preferenceScreen.preferenceCount)
                    .map { preferenceScreen.getPreference(it) }
                    .filterIsInstance<CheckBoxPreference>()
                    .apply {
                        forEach {
                            it.isChecked = !it.isChecked
                        }
                    }
        }

        private fun restartMode() {
            ImmersiveHandler.setMode(context, ImmersiveHandler.getMode(context))
        }

        companion object {

            fun newInstance(): SelectorFragment {
                return SelectorFragment()
            }
        }
    }
}
